working/333 - shared memory multithreading/proposal.md

```diff
@@ -0,0 +1,561 @@
+
+If you get a hold of `Zone` from another isolate (`Zone` should probably be `Shareable`) then you should be able to inject a callback into the event loop owned by that `Zone`.  
+
+#### Shareable `Future` and `Stream` instances
+
+`Future<Shareable>` and `Stream<Shareable>` should be shareable between isolates within the isolate group. 
+
+#### Structured Concurrency
+
+_Structured concurrency_ is a way of structuring concurrent code where lifecycle of concurrent tasks has clear relationship to the control-flow structure of the code which spawned those tasks. One of the most important properties of structured concurrency is an ability to cancel pending subtasks and propagate the cancellation recursively.
+
+Consider for example the following code:
+
+```dart
+Future<Result> doSomething() async {
+    final (a, b) = await (requestA(), computeB()).wait;
+    return combineIntoResult(a, b);
+}
+```
+
+If Dart supported _structured concurrency_, then the following would be guaranteed:
+
+* If either  `requestA`  or `computeB` fails, then the other is _canceled_.
+* `doSomething` computation can be _canceled_ by the holder of the `Future<Result>` and this cancellation will be propagated into `requestA` and `computeB`. 
+* If  `computeB` throws an error before `requestA` is awaited then `requestA` still gets properly canceled.
+
+Upgrading `dart:async` capabilities in the wake of shared-memory multithreading is also a good time to introduce some of the structured concurrency concepts into the language. See [Cancellable Future](https://gist.github.com/mraleph/6daf658c95be249c2f3cbf186a4205b9) proposal for the details of how this could work in Dart. 
+
+See also:
+
+* [JEP 453 - Structured Concurrency](https://openjdk.org/jeps/453)
+* [Wikipedia - Structured Concurrency](https://en.wikipedia.org/wiki/Structured_concurrency)
+
+
+
+### `dart:concurrent`
+
+`dart:concurrent` will serve as a library hosting low-level concurrency primitives. 
+
+#### `Thread` and `ThreadPool`
+
+Isolates are *not* threads even though they are often confused with ones. A code running within an isolate might be executing on a dedicated OS thread or it might running on a dedicated pool of threads. When expanding Dart's multithreading capabilities it seems reasonable to introduce more explicit ways to control threads. 
+
+```dart
+// dart:concurrent
+
+abstract class Thread implements Shareable {
+  /// Runs the given function in a new thread.
+  ///
+  /// The function is run in a shared isolate, meaning that 
+  /// it will not have access to the non-`@shared` state.
+  ///
+  /// The function will be run in a `Zone` which uses the
+  /// spawned thread as an executor for all callbacks: this
+  /// means the thread will remain alive as long as there is 
+  /// a callback referencing it.
+  external static Thread start<T>(FutureOr<T> Function() main); 
+  
+  external Thread get current;
```

Comment:
Does `Thread.current` always return a valid `Thread`, even for the root `Isolate`, and `Isolate`s created the old-fashioned way? Seems odd since status quo `Isolate`s can switch threads. If not, this method should probably return `Thread?`.

Reply:
I think it should return some thread always. It is okay that `Thread.current` can change between async suspensions (this will true for new fashioned multithreaded code as well - because you can run a thread pool based executor and have async code hop between threads as it is executed). 

I am going to add a comment here to clarify.